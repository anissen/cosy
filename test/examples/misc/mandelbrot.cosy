// Mandelbrot, adapted from https://github.com/wren-lang/wren/blob/main/example/mandelbrot.wren
foreign fn floor(v Num) Num

var yMin = -0.2
var yMax = 0.1
var xMin = -1.5
var xMax = -1.1

for yPixel in 0..24 {
    var y = (yPixel / 24) * (yMax - yMin) + yMin
    mut pixels = ''
    for xPixel in 0..80 {
        var x = (xPixel / 79) * (xMax - xMin) + xMin
        mut pixel = ' '
        mut x0 = x
        mut y0 = y
        for iter in 0..80 {
            mut x1 = (x0 * x0) - (y0 * y0)
            mut y1 = 2 * x0 * y0

            // Add the seed.
            x1 = x1 + x
            y1 = y1 + y

            x0 = x1
            y0 = y1

            // Stop if the point escaped.
            var d = (x0 * x0) + (y0 * y0)
            if (d > 4) {
                pixel = ' .:;+=xX$&'.split('')[floor(iter / 8)]
                break
            }
        }

        pixels += pixel
    }

    print pixels
}