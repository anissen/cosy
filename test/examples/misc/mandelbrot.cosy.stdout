                                         .................::::::x;+ x           
                                        ....::.......:::..:::  xx               
                                        ......:;:;;:;;x;;X::;;=                 
               .                       ...........::&+x    x=                   
               .......                ............:::+                          
               ...;........         ...............:::;;+&                      
               ....:;.........................;:.::=;;;                         
               ....::;.............;............:+xx                            
                .......:...........+............:::;+                           
                ........:;::..:...::+:..:;:...:::+                              
                ......:::+;:;:;:;::+ ;::X&::;:::;; $                            
               ..........:::;+ $=+ =    X   ;:::;;$                             
  ..............:......:::::;X                $;;+&                             
...:.............::.:..::;=                     =X                              
..................:;;::X;;                       X                              
....:.:...:....::::;;=    $                                                     
                                                                                
....:.:...:....::::;;=    $                                                     
..................:;;::X;;                       X                              
...:.............::.:..::;=                     =X                              
  ..............:......:::::;X                $;;+&                             
               ..........:::;+ $=+ =    X   ;:::;;$                             
                ......:::+;:;:;:;::+ ;::X&::;:::;; $                            
                ........:;::..:...::+:..:;:...:::+                              
